### DISCUSSION TITLE
Cannot read property of undefined (reading 'Role') sequence diagram jointjs

### DISCUSSION THREAD

**Discussion by AngelDavid** [clientIO/joint#2093] [Q&A]
### Introduction

Hello all,  Iam new to jointjs and I have created a react project and installed jointjs and also worked on er diagram.  Now I'm trying to use demo of sequence diagram into my project.  So the code/pages are as follows : 
In page1.jsx added the code available in the sequence.js file(available in the Demo Github link). In Return , , added the div element with id
Created a joint.shapes.sd.js in and added the code from the demo. 
Now when I run the page it shows the following error:
Cannot read property of undefined (reading 'Role') 
Any help in this would be great!! 

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by jamesgeorgewilliams** [comment]
It is difficult to say from the information provided here. If you want to create a JSFiddle or provide us with code, that would be a big help. The following link has some places where you can provide code online such as JSFiddle or Codepen.
https://github.com/clientIO/joint/wiki/Asking-Technical-Questions

One area I would pay attention to is that the namespaces are set up correctly in your project. You can learn more about custom shapes and namespaces in the following tutorial.
https://resources.jointjs.com/tutorial/ts-shape

**Reply by AngelDavid** [reply]
Hi @jamesgeorgewilliams the code looks like this:

home.jsx: 
import all the dependencies and jointjs
import '../joint.shapes.sd.js';
```
export default function App(){
(function(dia, sd, linkTools, elementTools, connectionStrategies, highlighters, paperElement) {

    const paperWidth = 800;
    const paperHeight = 600;
    const topY = 20;

    const graph = new dia.Graph();
    const paper = new dia.Paper({
        el: paperElement,
        width: paperWidth,
        height: paperHeight,
        model: graph,
        frozen: true,
        async: true,
        sorting: dia.Paper.sorting.APPROX,
        defaultConnectionPoint: { name: 'rectangle' },
        background: { color:  '#F3F7F6' },
        moveThreshold: 5,
        linkPinning: false,
        markAvailable: true,
        preventDefaultBlankAction: false,
        restrictTranslate: function(elementView) {
            const element = elementView.model;
            const padding = (element.isEmbedded()) ? 20 : 10;
            return {
                x: padding,
                y: element.getBBox().y,
                width: paperWidth - 2 * padding,
                height: 0
            };
        },
        interactive: function(cellView) {
            const cell = cellView.model;
            return (cell.isLink())
                ? { linkMove: false, labelMove: false }
                : true;
        },
        defaultLink: function(sourceView) {
            const type = sourceView.model.get('type');
            switch (type) {
                case 'sd.Message': {
                    return new sd.LifeSpan();
                }
                case 'sd.Lifeline': {
                    return new sd.Message();
                }
            }
            throw new Error('Unknown link type.');
        },
        connectionStrategy: function(end, cellView, magnet, p, link, endType, paper) {
            const type = link.get('type');
            switch (type) {
                case 'sd.LifeSpan': {
                    if (endType === 'source') {
                        end.anchor = { name: 'connectionRatio', args: { ratio: 1 }};
                    } else {
                        end.anchor = { name: 'connectionRatio', args: { ratio: 0 }};
                    }
                    return end;
                }
                case 'sd.Message': {
                    if (endType === 'source') {
                        return connectionStrategies.pinAbsolute.call(paper, end, cellView, magnet, p, link, endType, paper);
                    } else {
                        end.anchor = { name: 'connectionPerpendicular' };
                        return end;
                    }
                }
            }
        },
        validateConnection: function(cellViewS, magnetS, cellViewT, magnetT, end, linkView) {
            if (cellViewS === cellViewT) return false;
            const type = linkView.model.get('type');
            const targetType = cellViewT.model.get('type');
            switch (type) {
                case 'sd.Message': {
                    return targetType === 'sd.Lifeline';
                }
                case 'sd.LifeSpan': {
                    if (targetType !== 'sd.Message') return false;
                    if (cellViewT.model.source().id !== cellViewS.model.target().id) return false;
                    return true;
                }
            }
        },
        highlighting: {
            connecting: {
                name: 'addClass',
                options: {
                    className: 'highlighted-connecting'
                }
            }
        }
    });

    paper.el.style.border = '1px solid #E5E5E5';

    paper.on('link:pointermove', function(linkView, _evt, _x, y) {
        const type = linkView.model.get('type');
        switch (type) {
            case 'sd.Message': {
                const sView = linkView.sourceView;
                const tView = linkView.targetView;
                if (!sView || !tView) return;
                const padding = 20;
                const minY = Math.max(tView.sourcePoint.y - sView.sourcePoint.y, 0) + padding;
                const maxY = sView.targetPoint.y - sView.sourcePoint.y - padding;
                linkView.model.setStart(Math.min(Math.max(y - sView.sourcePoint.y, minY), maxY));
                break;
            }
            case 'sd.LifeSpan': {
                break;
            }
        }
    });

    paper.on('link:connect', function(linkView) {
        const type = linkView.model.get('type');
        switch (type) {
            case 'sd.Message': {
                editText(linkView, 'labels/0/attrs/labelText/text');
                break;
            }
            case 'sd.LifeSpan': {
                break;
            }
        }
    });

    const toolsScale = 1.2;

    graph.on('add', function(link) {
        if (!link.isLink()) return;
        const type = link.get('type');
        switch (type) {
            case 'sd.Lifeline': {
                const tools = new dia.ToolsView({
                    layer: null,
                    tools: [
                        new linkTools.HoverConnect({
                            scale: toolsScale
                        }),
                    ]
                });
                link.findView(paper).addTools(tools);
                break;
            }
        }
    });

    paper.on('cell:mouseenter', function(cellView) {
        const cell = cellView.model;
        const type = cell.get('type');
        switch (type) {
            case 'sd.Message': {
                const tools = new dia.ToolsView({
                    tools: [
                        new linkTools.Connect({
                            scale: toolsScale,
                            distance: -20
                        }),
                        new linkTools.Remove({
                            scale: toolsScale,
                            distance: 15
                        })
                    ]
                });
                cellView.addTools(tools);
                break;
            }
            case 'sd.LifeSpan': {
                const tools = new dia.ToolsView({
                    tools: [
                        new linkTools.Remove({
                            scale: toolsScale,
                            distance: 15
                        })
                    ]
                });
                cellView.addTools(tools);
                break;
            }
            case 'sd.Role': {
                const tools = new dia.ToolsView({
                    tools: [
                        new elementTools.Remove({
                            scale: toolsScale,
                            distance: '50%'
                        })
                    ]
                });
                cellView.addTools(tools);
                break;
            }
        }
    });

    paper.on('cell:mouseleave', function(cellView) {
        const cell = cellView.model;
        const type = cell.get('type');
        switch (type) {
            case 'sd.Role':
            case 'sd.LifeSpan':
            case 'sd.Message': {
                cellView.removeTools();
                break;
            }
        }
    });

    paper.on('blank:pointerdblclick', function(evt, x, y) {
        const role = new sd.Role({ position: { x: x - 50, y: topY }});
        role.addTo(graph);
        const lifeline = new sd.Lifeline();
        lifeline.attachToRole(role, paperHeight);
        lifeline.addTo(graph);
        editText(role.findView(paper), 'attrs/label/text');
    });

    const role1 = new sd.Role({ position: { x: 100, y: topY }});
    role1.setName('Browser');
    role1.addTo(graph);

    const role2 = new sd.Role({ position: { x: 400, y: topY }});
    role2.setName('Web Server');
    role2.addTo(graph);

    const role3 = new sd.Role({ position: { x: 600, y: topY }});
    role3.setName('Database Server');
    role3.addTo(graph);

    const backend = new sd.RoleGroup();
    backend.embed(role2);
    backend.embed(role3);
    backend.addTo(graph);
    backend.fitRoles();
    backend.listenTo(graph, 'change:position', function(cell) {
        if (cell.isEmbeddedIn(backend)) backend.fitRoles();
    });
    graph.on('remove', function(element) {
        if (!element.isElement()) return;
        const embeds = backend.getEmbeddedCells();
        if (embeds.length < 2) {
            backend.unembed(embeds);
            backend.remove();
        }
    });

    const lifeline1 = new sd.Lifeline();
    lifeline1.attachToRole(role1, paperHeight);
    lifeline1.addTo(graph);

    const lifeline2 = new sd.Lifeline();
    lifeline2.attachToRole(role2, paperHeight);
    lifeline2.addTo(graph);

    const lifeline3 = new sd.Lifeline();
    lifeline3.attachToRole(role3, paperHeight);
    lifeline3.addTo(graph);

    const message1 = new sd.Message();
    message1.setFromTo(lifeline1, lifeline2);
    message1.setStart(50);
    message1.setDescription('HTTP GET Request');
    message1.addTo(graph);

    const message2 = new sd.Message();
    message2.setFromTo(lifeline2, lifeline3);
    message2.setStart(150);
    message2.setDescription('SQL Command');
    message2.addTo(graph);

    const message3 = new sd.Message();
    message3.setFromTo(lifeline3, lifeline2);
    message3.setStart(250);
    message3.setDescription('Result Set');
    message3.addTo(graph);

    const message4 = new sd.Message();
    message4.setFromTo(lifeline2, lifeline1);
    message4.setStart(350);
    message4.setDescription('HTTP Response');
    message4.addTo(graph);

    const lifespan1 = new sd.LifeSpan();
    lifespan1.attachToMessages(message2, message3);
    lifespan1.addTo(graph);

    paper.unfreeze();

    // Text Editing

    function editText(cellView, textPath) {

        const cell = cellView.model;
        const textarea = document.createElement('textarea');
        textarea.style.position = 'absolute';
        textarea.style.width ='200px';
        textarea.style.height = '100px';
        textarea.style.left = '50%';
        textarea.style.top = `${paper.options.height / 2}px`;
        textarea.style.transform = 'translate(-50%, -50%)';
        textarea.style.padding = '5px';
        textarea.style.resize = 'none';
        textarea.style.boxShadow = '10px 10px 5px rgba(0, 0, 0, 0.5)';
        textarea.placeholder = cell.placeholder || 'Enter text here...';
        textarea.value = cell.prop(textPath) || '';
        document.body.appendChild(textarea);
        textarea.focus();
        textarea.setSelectionRange(0, textarea.value.length);

        cellView.paper.el.style.filter = 'blur(0.5px) grayscale(1)';
        cellView.paper.el.style.pointerEvents = 'none';

        const highlighter = highlighters.mask.add(cellView, 'root', 'selection', {
            layer: dia.Paper.Layers.FRONT,
            deep: true
        });

        function close() {
            textarea.remove();
            cellView.paper.el.style.filter = '';
            cellView.paper.el.style.pointerEvents = '';
            highlighter.remove();
        }

        function saveText() {
            cell.prop(textPath, textarea.value);
            close();
        }

        textarea.addEventListener('blur', saveText);

        textarea.addEventListener('keydown', function(evt) {
            if (evt.key === 'Enter' && !evt.shiftKey) {
                textarea.blur();
            }
            if (evt.key === 'Escape') {
                textarea.removeEventListener('blur', saveText);
                close();
            }
        });
    }

    paper.on('link:pointerdblclick', function(linkView, evt) {
        const labelIndex = linkView.findAttribute('label-idx', evt.target);
        if (!labelIndex) return;
        editText(linkView, `labels/${labelIndex}/attrs/labelText/text`);
    });

    paper.on('element:pointerdblclick', function(elementView, evt) {
        switch (elementView.model.get('type')) {
            case 'sd.Role': {
                editText(elementView, 'attrs/label/text');
                break;
            }
            case 'sd.RoleGroup': {
                editText(elementView, 'attrs/label/text');
                break;
            }
        }
    });

})(
    joint.dia,
    joint.shapes.sd,
    joint.linkTools,
    joint.elementTools,
    joint.connectionStrategies,
    joint.highlighters,
    document.getElementById('paper')
);
return{<div id="paper></div>}
```
joint.shapes.sd.js file
`(function(joint) {

    var standard = joint.shapes.standard;
    var dia = joint.dia;

    standard.Rectangle.define('sd.RoleGroup', {
        z: 1,
        attrs: {
            body: {
                stroke: '#DDDDDD',
                strokeWidth: 1,
                fill: '#F9FBFA'
            },
            label: {
                refY: null,
                refX: null,
                y: 'calc(h+2)',
                x: 'calc(w/2)',
                textAnchor: 'middle',
                textVerticalAnchor: 'top',
                fontSize: 12,
                fontFamily: 'sans-serif',
                textWrap: {
                    width: -10
                }
            }
        }
    }, {
        placeholder: 'What\'s the group\'s name?',

        fitRoles: function() {
            this.fitToChildren({ padding: 10 });
        }
    });

    standard.Rectangle.define('sd.Role', {
        z: 2,
        size: { width: 100, height: 80 },
        attrs: {
            body: {
                stroke: '#A0A0A0',
                strokeWidth: 1,
                rx: 2,
                ry: 2
            },
            label: {
                fontSize: 18,
                fontFamily: 'sans-serif',
                textWrap: {
                    width: -10
                }
            }
        }
    }, {
        placeholder: 'What\'s the role?',

        setName: function(name) {
            this.attr(['label', 'text'], name);
        }
    });

    standard.Link.define('sd.Lifeline', {
        z: 3,
        attrs: {
            line: {
                stroke: '#A0A0A0',
                strokeWidth: 1,
                strokeDasharray: '5,2',
                targetMarker: null
            }
        }
    }, {
        attachToRole: function(role, maxY) {
            const roleCenter = role.getBBox().center();
            this.set({
                source: { id: role.id },
                target: { x: roleCenter.x, y: maxY }
            });
            role.embed(this);
        }
    });

    dia.Link.define('sd.LifeSpan', {
        z: 4,
        attrs: {
            line: {
                connection: true,
                stroke: '#222222',
                strokeWidth: 2
            },
            wrapper: {
                connection: true
            },
            icon: {
                atConnectionRatioIgnoreGradient: 0.5
            }
        }
    }, {
        markup: [{
            tagName: 'path',
            selector: 'line',
            attributes: {
                'fill': 'none',
                'pointer-events': 'none'
            }
        }, {
            tagName: 'path',
            selector: 'wrapper',
            attributes: {
                'fill': 'none',
                'stroke': 'transparent',
                'stroke-width': 20
            }
        }, {
            tagName: 'g',
            selector: 'icon',
            children: [{
                tagName: 'circle',
                attributes: {
                    'r': 12,
                    'fill': '#222222'
                }
            }, {
                tagName: 'path',
                attributes: {
                    'd': 'M -3 -5 3 -5 3 -2 -3  2 -3 5 3 5 3 2 -3 -2 Z',
                    'stroke': '#FFFFFF',
                    'stroke-width': 1,
                    'fill': 'none'
                }
            }]
        }],
        attachToMessages: function(from, to) {
            this.source(from, { anchor: { name: 'connectionRatio', args: { ratio: 1 }}});
            this.target(to, { anchor: { name: 'connectionRatio', args: { ratio: 0 }}});
        }
    });

    standard.Link.define('sd.Message', {
        z: 5,
        source: { anchor: { name: 'connectionLength' }},
        target: { anchor: { name: 'connectionPerpendicular' }},
        attrs: {
            line: {
                stroke: '#4666E5',
                sourceMarker: {
                    'type': 'path',
                    'd': 'M -3 -3 -3 3 3 3 3 -3 z',
                    'stroke-width': 3
                }
            },
            wrapper: {
                strokeWidth: 20,
                cursor: 'grab'
            },
        }
    }, {
        placeholder: 'What\'s the message?',

        defaultLabel: {
            markup: [{
                tagName: 'rect',
                selector: 'labelBody'
            }, {
                tagName: 'text',
                selector: 'labelText'
            }],
            attrs: {
                labelBody: {
                    ref: 'labelText',
                    width: 'calc(w + 20)',
                    height: 'calc(h + 10)',
                    x: 'calc(x - 10)',
                    y: 'calc(y - 5)',
                    rx: 2,
                    ry: 2,
                    fill: '#4666E5'
                },
                labelText: {
                    fill: '#FFFFFF',
                    fontSize: 12,
                    fontFamily: 'sans-serif',
                    textAnchor: 'middle',
                    textVerticalAnchor: 'middle',
                    cursor: 'grab'
                }
            }
        },
        setStart: function(y) {
            this.prop(['source', 'anchor', 'args', 'length'], y);
        },
        setFromTo: function(from, to) {
            this.prop({
                source: { id: from.id },
                target: { id: to.id }
            });
        },
        setDescription: function(description) {
            this.labels([{ attrs: { labelText: { text: description }}}]);
        }
    });

})(joint);`

When i run the page, it shows couldn't read property of undefined(reading Role), the 'sd' is shown as undefined.

Kindly help me in this

**Reply by jamesgeorgewilliams** [reply]
It is a difficult to follow the code you provided. I would suggest updating the structure of your project to something like the following tutorial project. You will be able to use the same structure with `React` and `JointJS`.

https://resources.jointjs.com/tutorial/react-ts
https://github.com/clientIO/joint-plus-tutorial-react

If you still have some issues, you can create a JSFiddle with your updated project, and we can take a look.
https://jsfiddle.net/
